test/integration/deployment/deployment_test.go

```diff
@@ -633,3 +634,107 @@ func TestDeploymentLabelAdopted(t *testing.T) {
+
+	replicas := int32(1)
+	firstDeploymentName := "first-deployment"
+	secondDeploymentName := "second-deployment"
+	testers := []*deploymentTester{
+		{t: t, c: c, deployment: newDeployment(firstDeploymentName, ns.Name, replicas)},
+		{t: t, c: c, deployment: newDeployment(secondDeploymentName, ns.Name, replicas)},
+	}
+
+	// Start informer and controllers
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	go dc.Run(5, stopCh)
+
+	// Create 2 deployments with overlappings selectors
+	var err error
+	for _, tester := range testers {
+		tester.deployment, err = c.ExtensionsV1beta1().Deployments(ns.Name).Create(tester.deployment)
+		dname := tester.deployment.Name
+		if err != nil {
+			t.Fatalf("failed to create deployment %q: %v", dname, err)
+		}
+		// Wait for the deployment to be updated to revision 1
+		if err = tester.waitForDeploymentRevisionAndImage("1", fakeImage); err != nil {
+			t.Fatalf("failed to update deployment %q to revision 1: %v", dname, err)
+		}
+		// Make sure the deployment completes while manually marking its pods as ready at the same time
+		if err = tester.waitForDeploymentCompleteAndMarkPodsReady(); err != nil {
+			t.Fatalf("deployment %q failed to complete: %v", dname, err)
+		}
+	}
+
+	// Both deployments should proceed independently
+	rsSelector, err := metav1.LabelSelectorAsSelector(testers[0].deployment.Spec.Selector)
+	if err != nil {
+		t.Fatalf("failed to parse deployment selector: %v", err)
+	}
+	rsList, err := c.ExtensionsV1beta1().ReplicaSets(ns.Name).List(metav1.ListOptions{LabelSelector: rsSelector.String()})
+	if err != nil {
+		t.Fatalf("failed to list replicasets of deployments %q and %q in namespace %q: %v", firstDeploymentName, secondDeploymentName, ns.Name, err)
+	}
+	rss := rsList.Items
+	if len(rss) != 2 {
+		t.Fatalf("expected 2 replicasets to be found, but found %d", len(rss))
+	}
+
+	var mapping map[types.UID]*v1beta1.ReplicaSet
+	for _, rs := range rss {
+		controller := metav1.GetControllerOf(&rs)
+		if controller == nil {
+			t.Fatalf("replicaset %q is not owned by any deployment", rs.Name)
+		}
+
+		// Update [controllerRef's UID: replicaset] mapping for future usage
+		mapping[controller.UID] = &rs
+	}
+
+	// Verify replicasets are owned by the created deployments
+	for _, tester := range testers {
+		if _, ok := mapping[tester.deployment.UID]; !ok {
+			t.Fatalf("expected deployment %q owns replicaset %q or %q, but it does not", tester.deployment.Name, rss[0].Name, rss[1].Name)
+		}
+	}
+
+	// Scale only the first deployment by 1
+	newReplicas := replicas + 1
+	testers[0].deployment, err = testers[0].updateDeployment(func(update *v1beta1.Deployment) {
+		update.Spec.Replicas = &newReplicas
+	})
+	if err != nil {
+		t.Fatalf("failed updating deployment %q: %v", firstDeploymentName, err)
+	}
+
+	// Wait for the controller to notice the scale
+	if err := testers[0].waitForObservedDeployment(testers[0].deployment.Generation); err != nil {
```

Comment:
replace it with `waitForDeploymentComplete()`?

Reply:
ok

Reply:
As discussed, `waitForDeploymentComplete()` will not work. Let's use `waitForDeploymentCompleteAndMarkPodsReady`. I thought when scaling a deployment with 1 replicaset of 1 ready pod to 2 pods, the second pod will also have ready status. Spent some time to figure out that all pods have to be manually marked in integration test setting.